---
title: "02 - PreProcessing"
output: 
  html_document:
    css: "/local/workdir/zlr6/git_repos/biomi6300_moonmilk_amplicon_analysis/css/tufte.css"
    toc: yes
    toc_float:
      collapse: no
      smooth_scroll: yes
      toc_depth: 3
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
setwd("/local/workdir/zlr6/git_repos/biomi6300_moonmilk_amplicon_analysis")
knitr::opts_chunk$set(echo = TRUE, fig.align = "center", fig.path = "/local/workdir/zlr6/git_repos/biomi6300_moonmilk_amplicon_analysis/figures/02_PreProcessing/") # Send any figures to this folder
```

## Load libraries 
```{r load-libraries}
#devtools::install_github("joey711/phyloseq")
pacman::p_load(devtools, phyloseq, tidyverse, dada2, install = FALSE)
```

## Goals

Here, we will process the data into a phyloseq object. 

- ASV table 
- Taxonomy Table 
- Track Reads (metadata)

Then, we will remove the following: 

1. Remove Chloroplasts
2. Remove Mitochondria. 
3. Remove samples without "enough" reads. 

Finally, write data file of phyloseq output.


## Load Data 

### ASV Table

```{r load-asv-table}
# First, load asv table
load("/local/workdir/zlr6/git_repos/biomi6300_moonmilk_amplicon_analysis/data/01_DADA2/ASV_counts.RData")

# Inspect asv_tab
head(asv_tab)[,1:5]

# Fix names 
sample_titles <- colnames(asv_tab)
samples_fixed <- sapply(strsplit(basename(sample_titles), "_"), `[`,1) 
head(samples_fixed)

# re-write the ASV count file to fix names 
colnames(asv_tab) <- samples_fixed
str(asv_tab)
```

###  Taxonomy Table
```{r load-tax-table}
tax_df <- read.table("/local/workdir/zlr6/git_repos/biomi6300_moonmilk_amplicon_analysis/data/01_DADA2/ASV_taxonomy.tsv", sep = "\t", skip = 1)
head(tax_df)

# fix column names 
colnames(tax_df) <- c("asv_names", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "ASV", "ASVseq")

head(tax_df)

# Taxonomy Table Matrix
tax_mat <- 
  tax_df %>%
  tibble::column_to_rownames(., var = "asv_names") %>%
  as.matrix()
```

###  Track Reads Data
```{r load-track-reads}
load("/local/workdir/zlr6/git_repos/biomi6300_moonmilk_amplicon_analysis/data/01_DADA2/track_read_counts.RData")

# Take a look at the data
head(track_counts_df)
dim(track_counts_df)

# Load in metadata
metadata_df <- read.csv("/local/workdir/zlr6/git_repos/biomi6300_moonmilk_amplicon_analysis/data/02_PreProcessing/metadata.csv")
dim(metadata_df)
colnames(metadata_df)
# Replace the name of the column
colnames(metadata_df)[1] <- "names"

head(metadata_df)
head(track_counts_df)

metadata_track_reads_df <- 
  track_counts_df %>%
  left_join(., metadata_df, by = "names")

# Intuition check 
head(metadata_track_reads_df)

# Update row.names to be sample names 
## Before 
row.names(metadata_track_reads_df)
# Rewrite 
row.names(metadata_track_reads_df) <- metadata_track_reads_df$names
# Check afterwards that it worked 
row.names(metadata_track_reads_df)
# intuition check
head(metadata_track_reads_df)
```

## Handoff to phyloseq
```{r phyloseq-handoff}
# double check it's all good 
dim(asv_tab)
dim(tax_mat)

# Intuition check 
stopifnot(row.names(asv_tab) == row.names(tax_mat))

# Construct the phyloseq object 
raw_physeq <- phyloseq(otu_table(asv_tab, taxa_are_rows = TRUE),
                       sample_data(metadata_track_reads_df), tax_table(tax_mat))
                       
raw_physeq

# Save this raw phyloseq object 
save(raw_physeq, file = "data/02_PreProcessing/raw_physeq.RData")
```

## Clean up the data

Remove: 

1. Chloroplasts
2. mitochondria  

```{r rm-mitos-chloros}
# Remind myself of tax table 
#View(tax_mat)

# Make new physeq without chloroplasts
noChloros_physeq <- 
  raw_physeq %>% 
  # rm chloroplasts
  subset_taxa(Order != "Chloroplast" | is.na(Order))
  
# How many taxa were chloroplasts? 
num_chloro_ASVs <- ntaxa(raw_physeq) - ntaxa(noChloros_physeq)
num_chloro_ASVs

# Intuition chek 
#noChloros_physeq %>%
#  tax_table() %>%
#  data.frame() %>%
#  View()

# remove mitochondria 
noChlorosMitos_physeq <- 
  noChloros_physeq %>%
  subset_taxa(Family != "Mitochondria" | is.na(Family))

# How many mitochondrial ASVs? 
num_mito_ASVs <- ntaxa(noChloros_physeq) - ntaxa(noChlorosMitos_physeq)
num_mito_ASVs

noChlorosMitos_physeq

# How many total asvs were removed from chloros and mitos 
ntaxa(raw_physeq) - ntaxa(noChlorosMitos_physeq)
# proportion of asvs kept? 
ntaxa(noChlorosMitos_physeq)/ntaxa(raw_physeq)
```

# Evaulate and remove the control samples 

We do not have any control samples or a mock community.

# Evaluate the Sequencing Depth 

```{r seq-depth}
# The current data object
noChlorosMitos_physeq

# What is the library size/sequencing depth for each sample? 
seqSums_df <- 
  noChlorosMitos_physeq %>%
  otu_table() %>%
  # Sum each sample column 
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "names") %>%
  left_join(., metadata_track_reads_df, by = "names") 

# Rename second column 
colnames(seqSums_df)[2] <- "TotalSeqs"

# check
dim(seqSums_df)
head(seqSums_df)

# Show the depth of samples 
seqSums_df %>%
  dplyr::select(names, TotalSeqs) %>%
  arrange(TotalSeqs) %>%
  head()

# plot it! 
seqSums_df %>%
  ggplot(aes(x=reorder(names, TotalSeqs), y = TotalSeqs, fill = sample_title)) + 
  geom_bar(stat = "identity") 

# Density plot

seqSums_df %>%
  ggplot(aes(TotalSeqs, fill = sample_title)) +
  geom_density(alpha = 0.5)
```

## Session Information 
```{r session-info}
# Ensure reproducibility 
devtools::session_info()
```
