---
title: "01 - DADA2 Moonmilk"
output: 
  html_document:
    toc: yes
    toc_float:
      collapse: no
      smooth_scroll: yes
      toc_depth: 3
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align = "center", fig.path = "../figures/01_DADA2/") # Send any figures to this folder
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align = "center", fig.path = "../figures/01_DADA2/") # Send any figures to this folder
```

# Before you start

## Set my seed

```{r set-seed}
# Any number can be chosen
set.seed(3141596)
```

# Goals of this file

1. Use raw fastq files and generate quality plots to assess quality of reads.
2. Filter and trim out bad sequences and bases from our sequencing files.
3. Write out fastq files with high quality sequences.
4. Evaluate the quality from our filter and trim.
5. Infer errors on forward and reverse reads individually.
6. Identified ASVs on forward and reverse reads seperately using the error model.
7. Merge forward and reverse ASVs into "contiguous ASVs".
8. Generate the ASV count table (`otu_table` input for phyloseq).
9. Remove chimeras.
10. Track the read counts
11. Assign taxonomy.
12. Prepare the data for export!

Output that we need:

1. ASV Count Table: `otu_table`
2. Taxonomy Table: `tax_table`
3. Sample Information: `sample_data` - track the reads lost throughout the DADA2 workflow

